Sources/SwiftSyntax/SyntaxRawRewriter.swift.gyb

```diff
@@ -0,0 +1,279 @@
+  /// overflow during syntax tree rewriting.
+  /// 
+  /// To circumvent this problem, make calling the specific visitation function 
+  /// a two-step process: First determine the function to call in this function
+  /// and return a reference to it, then call it. This way, the stack frame
+  /// that determines the correct visitiation function will be popped of the
+  /// stack before the function is being called, making the switch's stack
+  /// space transient instead of having it linger in the call stack.
+  private func visitationFunc(for data: SyntaxData) -> ((SyntaxData) -> RawSyntax) {
+    switch data.raw.kind {
+    case .token:
+      return visitImplTokenSyntax
+    case .unknown:
+      return visitImplUnknownSyntax
+  % for node in NON_BASE_SYNTAX_NODES:
+    case .${node.swift_syntax_kind}:
+      return visitImpl${node.name}
+  % end
+    }
+  }
+
+  private func visit(_ data: SyntaxData) -> RawSyntax {
+    return visitationFunc(for: data)(data)
+  }
+
+#else
+
+  private func visit(_ data: SyntaxData) -> RawSyntax {
+    switch data.raw.kind {
+    case .token:
+      return visitImplTokenSyntax(data)
+    case .unknown:
+      return visitImplUnknownSyntax(data)
+  % for node in NON_BASE_SYNTAX_NODES:
+    case .${node.swift_syntax_kind}:
+      return visitImpl${node.name}(data)
+  % end
+    }
+  }
+
+#endif
+
+  private func visitChildren<SyntaxType: SyntaxProtocol>(
+    _ node: SyntaxType
+  ) -> RawSyntax {
+    // Walk over all children of this node and rewrite them. Don't store any 
+    // rewritten nodes until the first non-`nil` value is encountered. When this 
+    // happens, retrieve all previous syntax nodes from the parent node to 
+    // initialize the new layout. Once we know that we have to rewrite the 
+    // layout, we need to collect all futher children, regardless of whether 
+    // they are rewritten or not.
+    
+    // newLayout is nil until the first child node is rewritten and rewritten 
+    // nodes are being collected.
+    var newLayout: ContiguousArray<RawSyntax?>?
+
+    // Rewritten children just to keep their 'SyntaxArena' alive until they are
+    // wrapped with 'Syntax'
+    var rewrittens: ContiguousArray<RawSyntax> = []
+
+    let syntaxNode = node._syntaxNode
+
+    // Incrementing i manually is faster than using .enumerated()
+    var childIndex = 0
+    for (raw, info) in RawSyntaxChildren(syntaxNode) {
+      defer { childIndex += 1 }
+      guard let child = raw else {
+        // Node does not exist. If we are collecting rewritten nodes, we need to 
+        // collect this one as well, otherwise we can ignore it.
+        if newLayout != nil {
+          newLayout!.append(nil)
+        }
+        continue
+      }
+
+      // Build the Syntax node to rewrite
+      let absoluteRaw = AbsoluteRawSyntax(raw: child, info: info)
+      let data = SyntaxData(absoluteRaw, parent: syntaxNode)
+      
+      let rewritten = visit(data)
```

Comment:
Even at this point, we can't know the arena of `rewritten` is alive or not.
So it must be the rewriter (sub classes)'s responsibly to keep arena's alive until the rewriting completes.